src/AppInstallerCLICore/Workflows/UninstallFlow.cpp

```diff
@@ -0,0 +1,182 @@
+            break;
+        }
+        default:
+            THROW_HR(HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED));
+        }
+    }
+
+    void ExecuteUninstaller(Execution::Context& context)
+    {
+        const std::string installedTypeString = context.Get<Execution::Data::InstalledPackageVersion>()->GetMetadata()[PackageVersionMetadata::InstalledType];
+        switch (ManifestInstaller::ConvertToInstallerTypeEnum(installedTypeString))
+        {
+        case ManifestInstaller::InstallerTypeEnum::Exe:
+        case ManifestInstaller::InstallerTypeEnum::Burn:
+        case ManifestInstaller::InstallerTypeEnum::Inno:
+        case ManifestInstaller::InstallerTypeEnum::Msi:
+        case ManifestInstaller::InstallerTypeEnum::Nullsoft:
+        case ManifestInstaller::InstallerTypeEnum::Wix:
+            context << Workflow::ExecuteUninstallString;
+            break;
+        case ManifestInstaller::InstallerTypeEnum::Msix:
+            context << Workflow::MsixUninstall;
+            break;
+        case ManifestInstaller::InstallerTypeEnum::MSStore:
+            context <<
+                Workflow::EnsureFeatureEnabled(Settings::ExperimentalFeature::Feature::ExperimentalMSStore) <<
+                Workflow::MsixUninstall;
+            break;
+        default:
+        THROW_HR(HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED));
+        }
+    }
+
+    void ExecuteUninstallString(Execution::Context& context)
+    {
+        context.Reporter.Info() << Resource::String::UninstallFlowStartingPackageUninstall << std::endl;
+
+        auto uninstallResult = context.Reporter.ExecuteWithProgress(
+            std::bind(InvokeCreateProcess,
+                context.Get<Execution::Data::UninstallString>(),
+                std::placeholders::_1));
+
+        if (!uninstallResult)
+        {
+            context.Reporter.Warn() << Resource::String::UninstallAbandoned << std::endl;
+            AICLI_TERMINATE_CONTEXT(E_ABORT);
+        }
+        else if (uninstallResult.value() != 0)
+        {
+            // TODO: identify other success exit codes
+            context.Reporter.Error() << Resource::String::UninstallFailedWithCode << ' ' << uninstallResult.value() << std::endl;
+            AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_EXEC_UNINSTALL_COMMAND_FAILED);
+        }
+        else
+        {
+            context.Reporter.Info() << Resource::String::UninstallFlowUninstallSuccess << std::endl;
+        }
+    }
+
+    void MsixUninstall(Execution::Context& context)
+    {
+        const auto& packageFamilyNames = context.Get<Execution::Data::PackageFamilyNames>(); 
+        context.Reporter.Info() << Resource::String::UninstallFlowStartingPackageUninstall << std::endl;
+
+        for (const auto& packageFamilyName : packageFamilyNames)
+        {
+            auto packageFullName = Msix::GetPackageFullNameFromFamilyName(packageFamilyName);
+            if (packageFullName.has_value())
```

Comment:
>if (packageFullName.has_value()) [](start = 12, length = 32)

I understand it's less likely since this is from installed source. but do we want to log and show something to the user for "package not found"?

Reply:
We should at least AICLI_LOG that case, but I don't think it would be bad to output a warning as well.

---
In reply to: [536447224](https://github.com/Microsoft/winget-cli/pull/659#discussion_r536447224) [](ancestors = 536447224)